### DISCUSSION TITLE
Change Request: Allow disabling autofix for `reportUnusedDisableDirectives`

### DISCUSSION THREAD

**Discussion by cakoose** [eslint/eslint#17008] [Q&A] [answered]
### ESLint version

v8.29.0

### What problem do you want to solve?

In my ".eslintrc.js", I have set `reportUnusedDisableDirectives: true`, because  I want eslint to point out places where a disable directive is no longer needed. However, I _don't_ want eslint to autofix them.

It's not always 100% clear when an autofix is "safe". For example:
- I like eslint's formatting autofixes, like `semi`, `no-trailing-spaces`, etc.
- I like that eslint _doesn't_ autofix `no-unused-vars`.
- I disagree with eslint on `prefer-const`. I don't think it should autofix, but it does (https://github.com/eslint/eslint/issues/15943).

For `prefer-const`, I have a workaround: I use the [no-autofix plugin](https://github.com/aladdin-add/eslint-plugin-no-autofix). But that only works for _rules_, whereas `reportUnusedDisableDirectives` is not a rule.

### What do you think is the correct solution?

I'm not sure. Two:

**1. Make `reportUnusedDisableDirectives` a regular rule instead of a special option, which would allow me to use the `no-autofix` plugin.

Earlier discussion: https://github.com/eslint/eslint/issues/13104. It appears that there are other advantages to having it behave like a normal rule, but the linked issue makes it seem like it's not easy.

Maybe there's a way to have the _configuration_ look like a normal rule's configuration but still have the _implementation_ be something custom?

**2. Define two levels of auto-fix**

A "safe" level autofixes for things like `semi`, `no-trailing-spaces`, etc. This would be safe to execute more frequently, e.g. I use a keyboard shortcut on WebStorm, but some people configure them to run on save. I don't want `prefer-const` and `reportUnusedDisableDirectives` to change the code from underneath me while I'm still working on it.

A more aggressive level of autofixes for things like `prefer-const`, `reportUnusedDisableDirectives`. This would be useful when you need to apply a bunch of changes and have some other process where you actually review all the changes, e.g. via PR.

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by aladdin-add** [comment, marked as answer]
You are finding the option `--fix-type`: https://eslint.org/docs/latest/use/command-line-interface#--fix-type 😄

**Comment by cakoose** [comment]
Looks like `--fix-type` is what I'm looking for. Is there a way to specify it in ".eslintrc"? If not, I'll need to configure our IDEs (WebStorm and VSCode) to use `--fix-type layout`.

Tangent: @aladdin-add, are you the author of `eslint-plugin-no-autofix`? I'm using it just to turn off the auto-fix for the `prefer-const` rule, but with `--fix-type layout` that doesn't seem necessary anymore. Are there other specific examples where `eslint-plugin-no-autofix` is still useful?

**Comment by aladdin-add** [comment]
> If not, I'll need to configure our IDEs (WebStorm and VSCode) to use --fix-type layout.

I think so as it's only available in the cli.

re the `eslint-plugin-no-autofix`: you can safely remove it; but it's still useful when you want to turn off the fixer for certain layout rules.
